text/0328-polyglot-assert.md

```diff
@@ -0,0 +1,335 @@
+   ```
+
+### What are the mitigations and/or long term plans to address these concerns?
+
+Once this module reaches general availability, we will deprecate the old 'assert'
+module. At that point, all of its source code can now be made as part of the new
+'assert' module.
+
+The dependencies on 'cloudformation-diff' and 'cfnspec' will need to be carefully
+evaluated and a decision made on whether to lift-and-shift the relevant code, or
+to move these out of the monorepo as independent modules.
+
+Friendly interfaces for popular testing frameworks such as jest, nunit, junit, etc.
+can be built as separate modules specific to those languages (not jsii modules)
+on top of the new 'assert' library.
+If necessary, this can be owned and maintained by the CDK team, but probably more
+suited for the community to pick up and build as third party modules.
+
+### What alternative solutions did you consider?
+
+See [Appendix B](#appendix-b---alternatives) for alternatives.
+
+### What is the high level implementation plan?
+
+The implementation plan is fairly simple.
+
+- Implement and release the new assert module.
+- Migrate modules to use the new assert module.
+
+### Are there any open issues that need to be addressed later?
+
+None.
+
+## Appendix A - Detailed Design
+
+### Key Requirement
+
+The **key requirement** for the design choices is *dogfooding*.
+The new assert APIs must be usable from within the AWS CDK, and over time,
+all CDK modules that use the old assert module should be migrated to the new one.
+
+This is crucial to keep the assert library active and maintained.
+This is reinforced by the old assert library which, over time, has evolved with
+new features added to it.
+
+### Ship Vehicle
+
+The assert library works on `Stack` which is a construct available in `@aws-cdk/core`
+(or `aws-cdk-lib` in v2). Hence, takes a dependency on this construct.
+
+Given this and working backwards from the key requirement, the new assert module
+must be part of the AWS CDK monorepo.
+
+### Dependencies
+
+The main dependencies of the new assert module are the existing `assert` module,
+which in turn mainly depend on `cloudformation-diff` and `cfnspec`.
+The latter two are required to display human readable diffs, around what resources
+are different and how IAM permissions have changed.
+
+All three of these modules are pure typescript, and not jsii modules, whereas the
+new assert module is a jsii module.
+jsii only allows other jsii modules as regular dependencies, and any non jsii module
+should be [bundled](https://docs.npmjs.com/cli/v7/configuring-npm/package-json#bundleddependencies).
+
+However, all three modules are within the AWS CDK monorepo and the tooling of `yarn`,
+`lerna` and `npm` do not support bundling modules that are within the monorepo.
+
+For this reason, we have gone with the option to 'vendor in' these modules by copying
+the source directly into the new assert module with some minor adjustments.
```

Comment:
I am assuming copying will be done during build, correct? Or are we planning to copy & paste?

Reply:
Automatically done during build.